### PR TITLE
Fix documentation for \ and move it to linalg/generic.jl

### LINKED ISSUE #347: Matrix division fails when applied on a square matrix with duplicate rows (and columns)
I am trying to do least squares with a rank deficient matrix using  [the / operator](http://docs.julialang.org/en/release-0.4/stdlib/linalg/). It works in the general case, but fails when applied on a square symmetric matrix with duplicate rows (and columns).

``` julia
> n, p = 99, 8;
> y = randn(n);
> X = randn(n, p);
> X[:, 1] = 1;

# Rank deficient. X[:, 1] = X[:, 2] + X[:, 3] 
> X[:, 3] = 1 - X[:, 2];
> (X\y)'
1x8 Array{Float64,2}:
 -0.0643163  -0.123173  0.0588565  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931
> ((X'*X)\(X'*y))'
1x8 Array{Float64,2}:
 -0.124538  -0.0629509  0.119078  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931

# Duplicate column. X[:, 1] = X[:, 3] = 1
> X[:, 3] = 1;
> (X\y)'
1x8 Array{Float64,2}:
 -0.00272991  -0.182029  -0.00272991  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931
> (X'*X)\(X'*y)
ERROR: Base.LinAlg.SingularException(8)
 in \ at linalg/factorization.jl:25
 in \ at linalg/dense.jl:448
```

### LINKED ISSUE #17591: Matrix division fails when applied on a square matrix with duplicate rows (and columns)
I am trying to do least squares with a rank deficient matrix using  [the / operator](http://docs.julialang.org/en/release-0.4/stdlib/linalg/). It works in the general case, but fails when applied on a square symmetric matrix with duplicate rows (and columns).

``` julia
> n, p = 99, 8;
> y = randn(n);
> X = randn(n, p);
> X[:, 1] = 1;

# Rank deficient. X[:, 1] = X[:, 2] + X[:, 3] 
> X[:, 3] = 1 - X[:, 2];
> (X\y)'
1x8 Array{Float64,2}:
 -0.0643163  -0.123173  0.0588565  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931
> ((X'*X)\(X'*y))'
1x8 Array{Float64,2}:
 -0.124538  -0.0629509  0.119078  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931

# Duplicate column. X[:, 1] = X[:, 3] = 1
> X[:, 3] = 1;
> (X\y)'
1x8 Array{Float64,2}:
 -0.00272991  -0.182029  -0.00272991  0.0812868  -0.0363878  -0.0676099  -0.0715549  0.0175931
> (X'*X)\(X'*y)
ERROR: Base.LinAlg.SingularException(8)
 in \ at linalg/factorization.jl:25
 in \ at linalg/dense.jl:448
```

### PR DESCRIPTION
**Pull request by andreasnoack**
We don't use Bunch-Kaufman here.

Fixes JuliaLang/LinearAlgebra.jl#347

### CODE REVIEW

File: base/linalg/generic.jl [hunk 1]
```diff
@@ -329,6 +329,22 @@ function inv{T}(A::AbstractMatrix{T})
...
+"""
+    \\(A, B)
+
+Matrix division using a polyalgorithm. For input matrices `A` and `B`, the result `X` is
+such that `A*X == B` when `A` is square. The solver that is used depends upon the structure
+of `A`.  If `A` is upper or lower triangular (or diagonal), no factorization of `A` is
+required and the system is solved with either forward or backward substition.
+For non-triangular square matrices, an LU factorization is used.
+
+For rectangular `A` the result is the minimum-norm least squares solution computed by a
```
Comment (tkelman) [inline base/linalg/generic.jl]: whitespace error

File: base/linalg/generic.jl [hunk 2]
```diff
@@ -329,6 +329,22 @@ function inv{T}(A::AbstractMatrix{T})
     A_ldiv_B!(factorize(convert(AbstractMatrix{S}, A)), eye(S, checksquare(A)))
 end
 
+"""
+    \\(A, B)
+
+Matrix division using a polyalgorithm. For input matrices `A` and `B`, the result `X` is
+such that `A*X == B` when `A` is square. The solver that is used depends upon the structure
+of `A`.  If `A` is upper or lower triangular (or diagonal), no factorization of `A` is
+required and the system is solved with either forward or backward substition.
```
Comment (tkelman) [inline base/linalg/generic.jl]: substitution

**Comment by tkelman** [comment]
We aren't doing a symmetry check?

**Comment by andreasnoack** [comment]
No, we have never done that for `\`. I just went through the history to check and the documentation was wrongly changed long time ago. As I remember it, I decided not to use Bunch-Kaufman because it isn't significantly faster than LU and therefore the check wasn't worth it. This is in contrast to the `factorize` function which does check for symmetry and returns a Bunch-Kaufman factorization for non-positive definite symmetric matrices. For `factorize`, the output is a factorization so it makes sense to maintain the symmetric structure. Finally, we don't use `factorize` in `\` because `factorize` is type unstable by construction.

**Comment by tkelman** [comment]
Bunch-Kaufman should be about twice as fast as LU and use half as much memory in the n^3 terms.

**Comment by andreasnoack** [comment]
I don't think the memory access pattern for BK is as favorable as it is for LU. Just did some timings and it seems that the matrices have to be fairly large before BK is just slightly faster single threaded. On top of this, OpenBLAS has a multithreaded LU.

**Pull request merged**